### ISSUE TITLE
Investigate the issues with corpus merge timing out

### ISSUE THREAD

**Issue by Dor1s** [google/clusterfuzz#462]
#458 might help, but might not

The issue is that if at some point corpora for a target becomes too large, the merge / pruning can be always timing out since then. That means we'll be losing some efficiency and may not never resolve the issue.

**Comment by inferno-chromium** [comment]
I think we have a mitigation for this in corpus pruning task. Maybe the gsutil rsync itself fails due to large corpus size, need to check after 2 runs.

```
  # If our last execution failed, shrink to a randomized corpus of usable size
  # to prevent corpus from growing unbounded and recurring failures when trying
  # to minimize it.
  if last_execution_failed:
    _limit_corpus_size(context.initial_corpus_path,
                       CORPUS_SIZE_LIMIT_FOR_FAILURES)
```

**Comment by Dor1s** [comment]
Yeah, there might be different situations, e.g. `gsutil` timing out, and/or the target might be too slow and fails to merge even 10K inputs (as per the limit)?

I was thinking about another corpus pruning task (e.g. maybe weekly?) that would run on those problematic fuzz targets and allocate enough time to finish pruning for them.

Can't say that I like this idea, but that's what I'm doing locally from time to time when I have to fix such target.

**Comment by inferno-chromium** [comment]
corpus pruning task gets like a 24 hour duration every day (so i dont see advantage with weekly run). maybe just need to increase the gsutil duration for sync on those ? i am worried about local disk filing up too.

**Comment by Dor1s** [comment]
Oh, 24 hours is a lot! I forgot about that. Good point regarding `gsutil`, thanks! I should check stats + logs and see how many such targets we have and where they are failing.

**Comment by Dor1s** [comment]
Have been chasing one fuzz target affected, but turned out I hit https://bugs.chromium.org/p/chromium/issues/detail?id=952995 again, i.e. corpus pruning succeeded and we have all the stats, but https://clusterfuzz.com/fuzzer-stats/by-fuzzer/fuzzer/libFuzzer/job/libfuzzer_chrome_asan doesn't show those.

I'll check more fuzz targets.

**Comment by Dor1s** [comment]
Out of all Chromium fuzz targets, only 2 have problems with corpus pruning right now:

1. `grpc_uri_fuzzer_test` seems to be timing out, I'm still running it locally, will see how (if) it ends
2. `grpc_hpack_parser_fuzzer_test` quickly OOMs and then eventually ends up with a corrupted merge control file:

```
MERGE: failed to parse the control file (unexpected error)
```

which it cannot recover from.

So, I think I was wrong initially and our existing implementation does a great job and prunes everything that can be pruned.

I'll also skim through OSS-Fuzz targets and close this issue.

**Comment by Dor1s** [comment]
For `grpc_uri_fuzzer_test` the issue is `Failed to sync quarantine corpus to disk` (thanks @inferno-chromium for StackDriver kungfu), the quarantine has 3,384 files with a total size of 2.76 GB.

For `grpc_hpack_parser_fuzzer_test`, it's an OOMy target and an issue in libFuzzer: https://github.com/google/oss-fuzz/issues/2478.

**Comment by Dor1s** [comment]
We already have a 90 days lifecycle policy on the quarantine bucket. I was thinking to limit how much we can put into quarantine, but that seems to be a bad idea -- we can lose coverage if the target temporarily got broken.

Since we do not restore more than

```
MAX_QUARANTINE_UNITS_TO_RESTORE = 128
```

units per run, we may not sync the whole quarantine bucket to disk. But then it would change the logic of how we remove stuff from the quarantine bucket.

**Comment by Dor1s** [comment]
Ah, ok, I think we can copy just up to `MAX_QUARANTINE_UNITS_TO_RESTORE` units, delete them from the bucket, and after the pruning we will upload a new quarantine units back to GCS (just upload, not rsync). That way we won't lose anything and may even speed up some things when the bucket is unnecessarily huge.

**Comment by Dor1s** [comment]
Brainstormed this more with @inferno-chromium. The resolution is:
* Do not except from the pruning task when quarantine corpus fails to sync, just log an error (I'll upload a patch).
* If the quarantine bucket gets too large, we expect it to eventually free up by the lifecycle policy.
* Hacking some special logic for partial bucket downloading and removal doesn't seem worthwhile, given the low volume of such failures.

**Comment by Dor1s** [comment]
In OSS-Fuzz:

* 1 case when quarantine fails to sync -- should be addressed by #508 
* 5 cases when corpus fails to sync -- needs more thought in #507
* 4 error when there is a failure setting up a build, but those are not fatal, as I understand:

```
Error getting build url for job libfuzzer_asan_ffmpeg (r201905290220). 
Error getting build url for job libfuzzer_asan_dlplibs (r201905290206). 
Error getting build url for job libfuzzer_asan_imagemagick (r201905290144). 
Error getting build url for job libfuzzer_asan_ffmpeg (r201905290220). 
```

**Comment by inferno-chromium** [comment]
The fix for "Error getting build url" is https://github.com/google/clusterfuzz/pull/512
